### PR TITLE
Make themes dynamic and allow extensions to specify a module path

### LINKED ISSUE #2757: Make the extension metadata in package.json take a module for the plugin
Up to now, the extension metadata in the package.json only allows a plugin to export the plugin data as the default export of the package:
```json
  "jupyterlab": {
    "extension": true
  },
```
This means that any libraries the plugin imports must be available to npm. This is onerous if we are wrapping a small private dependency, or a dependency version we don't want to push to npm (a dev version, for example). In these cases, it would be much simpler to include the lab extension in the package itself - but that is impossible if the default export of the package *must* be the plugin data.

## Proposal
The `extension` attribute in package.json is either a boolean or a string. If it is `true`, look in the default exports of the package for the plugins as we do now. If it is a string, look in that file of the package for the plugins:
```json
  "jupyterlab": {
    "extension": "some/entry/point/that/exports/plugins"
  },
```
This makes it possible for a package to export some api *in addition* to providing a lab extension.

Another possibility is to make the string the *export* of the package to use (i.e., the export key name), but I think having a separate entry point may make more sense, as you may not want to export the lab plugin data in the public interface of the package.

### PR DESCRIPTION
**Pull request by blink1073**
Fixes #2757.  Partially addresses #1084.

- Allows extensions to specify a module path to load.
- Makes themes dynamically loaded and determined by a user setting.
- Adds documentation for settings and theme authoring.
- Dissolves `@jupyterlab/theming` into `@jupyterlab/application` and the two theme packages as appropriate.
- Adds more automation in the package integrity check and the creation of the `jupyterlab/package.json` file.

theme-switch

### CODE REVIEW

File: scripts/add-sibling.js [hunk 1]
```diff
@@ -66,10 +68,3 @@ var indexPath = path.join(basePath, 'packages', 'all-packages', 'src', 'index.ts
 var index = fs.readFileSync(indexPath, 'utf8');
 index = index + 'import "' + package.name + '";\n';
 fs.writeFileSync(indexPath, index);
-
-// Add the extension to jupyterlab/package.json
-var jupyterlabPackagePath = path.join(basePath, 'jupyterlab', 'package.json');
```
Comment (ian-r-rose) [inline scripts/add-sibling.js]: I confess to still being unclear on the intention of `jupyterlab/package.json` vs `jupyterlab/package.app.json`. If this is removed, does it need to be added elsewhere for the python package to pick up the new sibling?

In any event, a change here should be mirrored in `scripts/remove-sibling.js`.

Comment (blink1073) [inline scripts/add-sibling.js]: Thanks for the heads up.  `package.json` and `package.app.json` are mean for dev mode and app mode, respectively.  The latter is copy of `package.json` at the time of publishing the Python package.

Comment (ian-r-rose) [inline scripts/add-sibling.js]: So they are autogenerated (obviating the need to patch it in this script)?.

Comment (blink1073) [inline scripts/add-sibling.js]: `package.app.json` was already auto-generated, this PR adds auto-generation for `package.json` as well.

Comment (ian-r-rose) [inline scripts/add-sibling.js]: Ah, makes sense, thanks.

**Comment by blink1073** [comment]
Plan for handling of themes: 

- Add optional `themeCSS` metadata that is a path to a CSS file that is loaded as a string (not automatically added to the DOM).
- Add a new theme manager plugin has the following API:

```typescript
class ThemeManager
    // The themeCSS is populated in the webpack entry point
    constructor({ settingRegistry, themeCSS });
    theme: string;
    setTheme(name): Promise<void>;
    // Used for a theme picker
    themes: ReadonlyArray<string>;  
    register(theme: ITheme);
}

interface ITheme {
    name: string;
    displayName: string;
    load(): Promise<void>;
    unload(): Promise<void>
}
```

- A theme is loaded by 
  - Unloading the previous theme (if applicable), and removing its CSS from the private style tag
  - Adding the new CSS to the private style tag, and then loading the new theme
- The current theme is linked to a setting in the theme manager plugin

In this way, we meet the following goals:
- The initial theme CSS can be loaded synchronously during app startup, preventing a jarring change in fonts/colors
- Themes are discoverable and can be changed dynamically without page reload
- The theme author does not need to use any WebPack-specific syntax

**Comment by blink1073** [comment]
We can't switch themes dynamically because we can't load additional CSS that is 
removable without exposing the loading pattern (`let themeCSS = require('!css-loader!foo.css').toString()`).

So, a themeExtension is a standard application extension, but has the following metadata that allows us to discover and load only the appropriate theme plugin:

```
jupyterlab: {
    themeExtension: {
       name: 'JupyterLab Dark',
       path?: ''
    }
}
```

**Comment by blink1073** [comment]
A plan that allows for dynamic loading:

- Add a `/themes` folder to the application that has the static theme assets, copied when the extension is installed.
- Add an `api/themes` endpoint to serve these files.
- The client side theme manager has a `loadCSS(path: string)` method that makes a request to the `api/themes` endpoint for the asset and adds it to a `script` tag.  It maintains a disposable set that removes any script tags added by the current theme.
- We can move the light and dark theme icons to their respective themes, and they will not be a part of the main bundle.

The downside is that any assets used from third party libs must be included in the package (e.g. icon files), since we will not have WebPack to resolve the dependency paths.

**Comment by blink1073** [comment]
Proof of concept of the above done, loading a theme dynamically using a `link` tag.  The only additional change made is that relative `url()` specifiers in CSS need to include the appropriate static link, e.g.: `  --jp-image-jupyter: url('./lab/api/themes/jupyterlab-theme-light-extension/style/images/jupyter.svg');`, otherwise they are loaded relative to the page url.

**Comment by blink1073** [comment]
Progress:

theme-switch

**Comment by blink1073** [comment]
cc @cameronoelsen @ellisonbg

File: packages/theme-light-extension/src/index.ts [hunk 2]
```diff
@@ -1,19 +1,31 @@
...
- */
-const extension: JupyterLabPlugin<void> = {
-  id: 'jupyter.themes.light',
-  autoStart: true,
-  activate: (app) => {
-    // No-op.
-  }
+
+const plugin: JupyterLabPlugin<void> = {
+  id: 'theme-light-extension',
```
Comment (afshin) [inline packages/theme-light-extension/src/index.ts]: What is the rationale for not calling these `jupyter.themes.light` or 'jupyter.extensions.light-theme' or some variety of `jupyter.foo.bar`?

**Comment by santiagobasulto** [comment]
Hello people. I was reading this PR trying to understand where can I set the default theme for all my sessions. Is there any file I could write in my $HOME that will read automatically the settings? So I can leave the Dark template set by default every time I run the lab? Thanks!

**Comment by pisymbol** [comment]
Is there any configuration file that would allow me to start jupyterlab with the dark theme as the default? This seems like an obvious and good feature to have.

**Comment by vidartf** [comment]
@pisymbol Thanks for your feedback! Please open a new issue about this, to increase visibility for the feature request, and to ensure it doesn't simply disappear (comments on merged PRs will not be picked up during triage, and it doesn't make sense to reopen a PR).

**Comment by vidartf** [comment]
PS: If your reason for not directly opening an issue was that you were unsure if such functionality already existed, discourse might be a good alternative: https://discourse.jupyter.org/c/jupyterlab

**Comment by jasongrout** [comment]
If you change the theme, it should be saved and start with that theme. Is that what you are asking?

**Comment by pisymbol** [comment]
I'm saying I should be able to write out a json config parameter somewhere that specifies what theme I I want on bootup (the default theme - in this case it maybe just light or dark but I can imagine people will want to specify their own theme).

Think about for example setting up a docker container with JuptyerLab in it and in the Dockerfile I want to be able to autoconfigure the dark theme as my default theme. Optionally, there could also be a command line argument to change the initial theme for the current session. Best of both worlds.

Does that make sense?

**Comment by ian-r-rose** [comment]
@pisymbol  you can distribute a settings JSON file that specifies the default theme with your docker container. The last-set theme is already saved in the settings system, so that should already persist between launches of the application.
An example of this can be found here: https://github.com/ian-r-rose/binder-workspace-demo

**Comment by jasongrout** [comment]
And there is actually one more place settings can be overridden, between the defaults specified in the plugin and the user settings: the application-level default settings. That may be a more appropriate place for overriding settings without touching the user settings files.

The directory for this setting override is configured in the lab app:

**Pull request merged**